components/observatorium.libsonnet

```diff
@@ -0,0 +1,144 @@
+          volumeClaimTemplate: error 'must provide volumeClaimTemplate',
+          replicas: 3,
+          replicationFactor: 3,
+          retention: '6h',
+          hashringConfigMapName: '%s-generated' % obs.thanosReceiveController.configmap.metadata.name,
+        },
+        statefulSet+: {
+          metadata+: {
+            labels+: {
+              'controller.receive.thanos.io/hashring': hashring.hashring,
+              'controller.receive.thanos.io': 'thanos-receive-controller',
+            },
+          },
+        },
+      }
+    for hashring in obs.config.hashrings
+  },
+
+  rule:: t.rule + t.rule.withVolumeClaimTemplate + {
+    config+:: {
+      name: 'thanos-rule',
+      namespace: obs.config.namespace,
+      replicas: 2,
+    },
+  },
+
+  store:: t.store + t.store.withVolumeClaimTemplate + {
+    config+:: {
+      name: 'thanos-store',
+      namespace: obs.config.namespace,
+      replicas: 1,
+    },
+  },
+
+  query:: t.query {
+    config+:: {
+      name: 'thanos-query',
+      namespace: obs.config.namespace,
+      replicas: 1,
+      stores: [
+        'dnssrv+_grpc._tcp.%s.%s.svc.cluster.local' % [service.metadata.name, service.metadata.namespace]
+        for service in [obs.rule.service, obs.store.service] + [obs.receivers[hashring].service for hashring in std.objectFields(obs.receivers)]
+      ],
+      replicaLabels: ['prometheus_replica', 'ruler_replica', 'replica'],
+    },
+  },
+
+  queryCache:: (import 'cortex-query-frontend.libsonnet') + {
+    config+:: {
+      name: 'cortex-query-cache',
+      namespace: obs.config.namespace,
+      downstreamURL: 'http://%s.%s.svc.cluster.local.:%d' % [
+        obs.query.service.metadata.name,
+        obs.query.service.metadata.namespace,
+        obs.query.service.spec.ports[1].port,
+      ],
+    },
+  },
+} + {
+  local obs = self,
+
+  rule+:: {
+    config+:: {
+      queriers: ['dnssrv+_http._tcp.%s.%s.svc.cluster.local' % [obs.query.service.metadata.name, obs.query.service.metadata.namespace]],
+    },
+  },
+
+  manifests+:: {
+    ['thanos-query-' + name]: obs.query[name]
+    for name in std.objectFields(obs.query)
+  } + {
+    ['query-cache-' + name]: obs.queryCache[name]
+    for name in std.objectFields(obs.queryCache)
+  } + {
+    ['thanos-receive-' + hashring + '-' + name]: obs.receivers[hashring][name]
+    for hashring in std.objectFields(obs.receivers)
+    for name in std.objectFields(obs.receivers[hashring])
+  } + {
+    thanosReceiveService:
+      local service = k.core.v1.service;
+      local ports = service.mixin.spec.portsType;
+
+      service.new(
+        'thanos-receive',
+        { 'app.kubernetes.io/component': 'thanos-receive' },
+        [
+          ports.newNamed('grpc', 10901, 10901),
+          ports.newNamed('http', 10902, 10902),
+          ports.newNamed('remote-write', 19291, 19291),
+        ]
+      ) +
+      service.mixin.metadata.withNamespace(obs.config.namespace),
```

Comment:
It would still be better to move this out of here, to have a simple looping over things to get manifests. Why doesn't it fit into the other component?

Reply:
Yeah I've thought about that as well, but I'm unsure where to put it .. this is essentially a "system global object that we only need once" which doesn't really align with the rest of the receive components. If you have a suggestion I'm all ears. I'm also happy to start a "global" section, wdyt?

Reply:
I would have said that we should have somehow put it in the same obs.receivers object, but then discovered that that is nested 2 levels. Putting it into the top level one would then break the level 2 for-loop. 
Other than leaving it here (which might just want end up happening then) I could think of having another `obs.receiversService` for the sake of having 
```
{
    ['thanos-receive-' + hashring + '-' + name]: obs.receivers[hashring][name]
    for hashring in std.objectFields(obs.receivers)
    for name in std.objectFields(obs.receivers[hashring])
}  + {
    ['thanos-receive-' + name]: obs.receiversService[name]
    for name in std.objectFields(obs.receiversService)
}
```

Reply:
At the same time, having the 2 level nested component feels a bit like a layer violation. Maybe we can think of a different approach to have the `obs.receiver` still expose all objects in the top level? The nesting of multiple receive hashrings seems more like a implementation detail from a component point of view. :thinking:

Reply:
Why do you say two levels? I see only the level of where we loop per hashring, which is pretty much exactly what we want. I wanted to avoid flattening the objects immediately as this way we can treat each hashring and patch each hashring individually and the flattening is really just a thing to output. In reality the objects should be nested I think.